hazelcast/src/test/java/com/hazelcast/map/listener/EntryUpdatedListenerTest.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.listener;
+
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.map.IMap;
+import com.hazelcast.map.MapInterceptor;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class EntryUpdatedListenerTest extends HazelcastTestSupport {
+    /**
+     * @see <a href="https://hazelcast.atlassian.net/browse/HZ-2837">HZ-2837 - Field level mutation being taken by listener as
+     *      old value but not being considered by interceptor - Strange Behaviour</a>
+     */
+    @Test
+    public void testOldValues() throws InterruptedException, ExecutionException {
+        final HazelcastInstance instance = createHazelcastInstanceFactory().newHazelcastInstance();
+
+        // Create a map
+        final IMap<Object, AtomicInteger> map = instance.getMap(randomMapName());
+
+        // Setup the listeners
+        final CompletableFuture<Integer> entryListenerOldValue = setMapListener(
+                listener -> map.addEntryListener(listener, true));
+        final CompletableFuture<Integer> entryLocalListenerOldValue = setMapListener(map::addLocalEntryListener);
+
+        final CompletableFuture<Integer> interceptorOldValue = new CompletableFuture<>();
+        map.addInterceptor(new MapInterceptor() {
```

Comment:
Maybe you can create a `MapInterceptorAdaptor` and override the needed methods to increase readability.